Agents/FenlandTrajectoryAgent/agent/exposure_calculator/exposure_utilities.py

```diff
@@ -0,0 +1,1023 @@
+        finally:
+            if conn:
+                conn.close()
+
+        df = df.rename(columns={
+            "column1": "SPEED",
+            "column2": "DISTANCE",
+            "column3": "HEIGHT",
+            "column4": "HEADING",
+            "column5": "LATITUDE",
+            "column6": "LONGITUDE",
+            "column7": "POINT"
+        })
+
+        if 'time' in df.columns:
+            df['time'] = pd.to_datetime(df['time'], errors='coerce')
+            df = df.sort_values('time').reset_index(drop=True)
+
+        # Find earliest time, store to df['trajectory_start_time']
+        if not df.empty:
+            earliest_dt = df['time'].iloc[0]
+            earliest_str = earliest_dt.strftime('%Y-%m-%dT%H:%M:%SZ')
+            df['trajectory_start_time'] = earliest_str
+            self.logger.info(f"Trajectory earliest time = {earliest_str}")
+        else:
+            self.logger.warning("Trajectory DataFrame is empty; no earliest time")
+
+        return df
+
+    def get_domain_and_featuretype(self, env_data_iri: str, endpoint_url: Optional[str] = None) -> Tuple[str, str]:
+        if endpoint_url is None:
+            endpoint_url = self.ENV_DATA_ENDPOINT_URL
+        template = self.load_template("get_domain_and_featuretype.sparql")
+        query = template.format(env_data_iri=env_data_iri)
+        headers = {"Content-Type": "application/sparql-query", "Accept": "application/json"}
+        resp = requests.post(endpoint_url, data=query, headers=headers, timeout=30)
+        resp.raise_for_status()
+        results = resp.json().get("results", {}).get("bindings", [])
+        if not results:
+            raise ValueError(f"No domainName/featureType found for {env_data_iri}")
+        domain_name = results[0]["domainName"]["value"]
+        feature_type = results[0]["featureType"]["value"]
+        self.logger.info(f"For {env_data_iri}, domainName={domain_name}, featureType={feature_type}")
+        return domain_name, feature_type
+
+    # ---- NEW OR MODIFIED ----
+    # => Here we use absolute-time-filter.sparql
+    def fetch_env_data(self, env_data_iri: str, endpoint_url: Optional[str] = None,
+                       reference_time: Optional[str] = None) -> Tuple[pd.DataFrame, str]:
+        if endpoint_url is None:
+            endpoint_url = self.ENV_DATA_ENDPOINT_URL
+
+        domain_label, feature_type = self.get_domain_and_featuretype(env_data_iri, endpoint_url)
+        dl_lower = domain_label.strip().lower()
+
+        self.logger.info(f"[fetch_env_data] Processing env_data_iri={env_data_iri}, domain_label={domain_label}, feature_type={feature_type}")
+
+        if env_data_iri in self.env_data_cache and reference_time is None:
+            self.logger.info(f"[fetch_env_data] Cache hit for env_data_iri: {env_data_iri}")
+            return self.env_data_cache[env_data_iri]
+
+        if feature_type.upper() == "POINT" and dl_lower.startswith("food hygiene rating") and reference_time is not None:
```

Comment:
I think this entire if-else block can be made more concise and explained more clearly via comments or in the readme. E.g. there are many things that can be retrieved via a single query, rather than choosing the appropriate query via different if-else blocks.

This avoids issues such as get_point_generic being out-of-date relative to absolute_time_filter, which is the case at the moment